ark/analysis/visualize.py

```diff
@@ -1,9 +1,65 @@
+import os
 import pandas as pd
 import numpy as np
 import matplotlib.pyplot as plt
 import seaborn as sns
 
 
+def draw_boxplot(cell_data, col_name, col_split=None, split_vals=None, save_dir=None):
+    """Draws a boxplot for a given column, optionally with help from a split column
+
+    Args:
+        cell_data (pandas.DataFrame):
+            Dataframe containing columns with Patient ID and Cell Name
+        col_name (str):
+            Name of the column we wish to draw a box-and-whisker plot for
+        col_split (str):
+            If specified, used for additional box-and-whisker plot faceting
+        split_vals (list):
+            If specified, only visualize the specified values in the col_split column
+        save_dir (str):
+            If specified, a directory where we will save the plot
+    """
+
+    # the col_name must be valid
+    if col_name not in cell_data.columns.values:
+        raise ValueError("col_name specified does not exist in data provided")
+
+    # basic error checks if split_vals is set
+    if split_vals is not None:
+        # the user cannot specify split_vales without specifying col_split
+        if col_split is None:
+            raise ValueError("If split_vals is set, then col_split must also be set")
+
+        # all the values in split_vals must exist in the col_name of cell_data
+        if not all(val in cell_data[col_split].unique() for val in split_vals):
+            raise ValueError("Some values in split_vals not found in col_split column of data")
+
+    # don't modify cell_data in anyway
+    data_to_viz = cell_data.copy(deep=True)
+
+    # ignore values in col_split not in split_vals if split_vals is set
+    if split_vals:
+        data_to_viz = data_to_viz[data_to_viz[col_split].isin(split_vals)]
+
+    if col_split:
+        # if col_split, then we explicitly facet the visualization
+        # labels are automatically generated in Seaborn
+        sns.boxplot(x=col_split, y=col_name, data=cell_data)
+        plt.title("Distribution of %s, faceted by %s" % (col_name, col_split))
+    else:
+        # otherwise, we don't facet anything, but we have to explicitly make vertical
+        sns.boxplot(x=col_name, data=cell_data, orient="v")
+        plt.title("Distribution of %s" % col_name)
+
+    # save visualization to a directory if specified
+    if save_dir is not None:
+        if not os.path.exists(save_dir):
+            raise ValueError("save_dir %s does not exist" % save_dir)
+
+        plt.savefig(os.path.join(save_dir, "sample_boxplot_viz.png"))
```

Comment:
@ngreenwald in general for the visualization plots, if `save_dir` doesn't exist, should we create that directory by default, provide an option to do so, or just throw an error as above?  

I know it's easy to externally create a folder, but that mandates that users do some tinkering around.

Reply:
Looking at `example_visualization.ipynb`, the run will actually fail if you don't have a visualization folder specified in data/example_dataset. We can do one of two things:

- Create the directory if it does not exist
- Push an example visualization folder to the repo (not a huge fan of this option, but if it makes it easier to do then I wouldn't mind)

Reply:
I think this function is low level enough that it shouldn't create its own directory. whatever is calling this should handle the folder creation. For example the first cell in the notebook

Reply:
@ngreenwald OK, then would it be best to add a command in the notebook to create `data/example_dataset/visualization` if it doesn't exist?